volatility/framework/plugins/windows/netlist.py

```diff
@@ -0,0 +1,544 @@
+        # first decode the pointer
+        netw_inside = cls._decode_pointer(assignment.Entry)
+
+        if netw_inside:
+            # if the value is valid, calculate the actual object address by subtracting the offset
+            curr_obj = context.object(obj_name, layer_name = layer_name, offset = netw_inside - ptr_offset)
+            vollog.debug("Found {} object @ 0x{:2x}, yielding...".format(proto, curr_obj.vol.offset))
+            yield curr_obj
+
+            # if the same port is used on different interfaces multiple objects are created
+            # those can be found by following the pointer within the object's `Next` field until it is empty
+            while curr_obj.Next:
+                curr_obj = context.object(obj_name, layer_name = layer_name, offset = cls._decode_pointer(curr_obj.Next) - ptr_offset)
+                yield curr_obj
+
+    @classmethod
+    def get_tcpip_module(cls,
+                         context: interfaces.context.ContextInterface,
+                         layer_name: str,
+                         nt_symbols: str) -> interfaces.objects.ObjectInterface:
+        """Uses `windows.modules` to find tcpip.sys in memory.
+
+        Args:
+            context: The context to retrieve required elements (layers, symbol tables) from
+            layer_name: The name of the layer on which to operate
+            nt_symbols: The name of the table containing the kernel symbols
+
+        Returns:
+            The constructed tcpip.sys module object.
+        """
+        for mod in modules.Modules.list_modules(context, layer_name, nt_symbols):
+            if mod.BaseDllName.get_string() == "tcpip.sys":
+                vollog.debug("Found tcpip.sys image base @ 0x{:x}".format(mod.DllBase))
+                return mod
+
+    @classmethod
+    def parse_hashtable(cls,
+                        context: interfaces.context.ContextInterface,
+                        layer_name: str,
+                        ht_offset: int,
+                        ht_length: int,
+                        alignment: int,
+                        pointer_length: int) -> list:
+        """Parses a hashtable quick and dirty.
+
+        Args:
+            context: The context to retrieve required elements (layers, symbol tables) from
+            layer_name: The name of the layer on which to operate
+            ht_offset: Beginning of the hash table
+            ht_length: Length of the hash table
+            pointer_length: Length of this architecture's pointers
+
+        Returns:
+            The hash table entries which are _not_ empty
+        """
+        for index in range(ht_length):
+            # mask pointer so we do not get confused with abbreviated virtual offsets
+            # this currently only works for 64-bit.
+            # TODO: add x86 support.
+            current_qword = (0xffff000000000000 | cls.read_pointer(context, layer_name, ht_offset + index * alignment, pointer_length))
```

Comment:
Well, let's approach it from this point. The use case is to parse a hash table and find its values. Entries of the hash table which are valid can be detected by them being not pointers to themselves, but different ones. This is why I hacked together this function here.

Just based upon this requirement, maybe we can find a better way to parse the table without hardcoded masks? :D

Reply:
I dove into this again. What we have here is a structure of type `_RTL_DYNAMIC_HASH_TABLE`'s `Directory` entries. The struct is defined as follows: http://librecrops.github.io/lost-sdk/files/RTL_DYNAMIC_HASH_TABLE.h.html

When we follow the `Directory` pointer we see the following in memory:
```
(primary) >>> dq(rtl_dyn_hash_table.Directory)
0xc88d39c81800    ffffc88d39c81800 ffffc88d39c81800    ....9... ....9...
0xc88d39c81810    ffffc88d39c81810 ffffc88d39c81810    ....9... ....9...
0xc88d39c81820    ffffc88d39c81820 ffffc88d39c81820    ....9... ....9...
0xc88d39c81830    ffffc88d39c81830 ffffc88d39c81830    ....9..0 ....9..0
0xc88d39c81840    ffffc88d39c81840 ffffc88d39c81840    ....9..@ ....9..@
0xc88d39c81850    ffffc88d39c81850 ffffc88d39c81850    ....9..P ....9..P
0xc88d39c81860    ffffc88d39c81860 ffffc88d39c81860    ....9.. ....9..
<..>
0xc88d39c81950    ffffc88d39c81950 ffffc88d39c81950    ....9..P ....9..P
0xc88d39c81960    ffffc88d3d649228 ffffc88d3d649228    ....=d.( ....=d.(
0xc88d39c81970    ffffc88d39c81970 ffffc88d39c81970    ....9..p ....9..p
```
and we are looking for values like at `0xc88d39c81960` which are not their own address. Unfortunately if we look at the definition above there is no defined data structure for those entries. While every instance of them looks like a `_LIST_ENTRY` there is no NT-symbols-defined struct for the shown list.  ( There are some nt structs which look like they would fit, for instance `_RTL_DYNAMIC_HASH_TABLE_ENTRY` (size of 24 byte, does not fit) or `_SINGLE_LIST_ENTRY` (size of 8 bytes, would fit twice) but I don't think they help much given that we would require a array-like struct for easier parsing...)

I see two options: either parse that list manually like I do it right now or define a custom "list of `_LIST_ENTRY`" in my netscan symbol tables so we can benefit from... a little bit more readability for a lot more effort?

I tend towards option 1. Or maybe I'm overlooking something :D

Reply:
Parsing it manually is fine, but given it's an address, you can cast it as a pointer and it should mask it to the appropriate size of the address space.  It may not be worthwhile, but hardcoded things will inevitably need changing, so if we can just cast those offsets as pointers (possibly to a known structure, but even if not), then the framework should handle the bulk of it for you?  That doesn't require hand coding a new full structure, but should give us all the benefits of letting the framework look after all the details.  Does that sound workable?  5:)

Reply:
I've reworked it into comparing `pointer == pointer.vol.offset` which works. Is this stable enough or is there a prettier way?
```python
        for index in range(ht_length):
            current_addr = ht_offset + index * alignment
            current_pointer = context.object(net_symbol_table + constants.BANG + "pointer",
                                    layer_name = layer_name,
                                    offset = current_addr)
            # check if addr of pointer is equal to the value pointed to
            if current_pointer.vol.offset == current_pointer:
                continue
            yield current_pointer
```

Reply:
Yep, that looks fine to me.  5:)